pkg/storage/stores/shipper/bloomshipper/store.go

```diff
@@ -156,17 +162,49 @@ func (b *bloomStoreEntry) TenantFilesForInterval(ctx context.Context, interval I
 			continue
 		}
 
-		for _, object := range objects {
-			tenant, err := b.ParseTenantKey(key(object.Key))
+		// Sort objects by the key to ensure keys are sorted by tenant.
+		cmpObj := func(a, b client.StorageObject) int {
+			if a.Key < b.Key {
+				return -1
+			}
+			if a.Key > b.Key {
+				return 1
+			}
+			return 0
+		}
+		if !slices.IsSortedFunc(objects, cmpObj) {
+			slices.SortFunc(objects, cmpObj)
+		}
+
+		for i := 0; i < len(objects); i++ {
+			tenant, err := b.TenantPrefix(key(objects[i].Key))
 			if err != nil {
-				return nil, fmt.Errorf("error parsing tenant key [%s]: %w", object.Key, err)
+				return nil, fmt.Errorf("error parsing tenant key [%s]: %w", objects[i].Key, err)
+			}
+
+			// Search next object with different tenant
+			var j int
+			for j = i + 1; j < len(objects); j++ {
+				nextTenant, err := b.TenantPrefix(key(objects[j].Key))
+				if err != nil {
+					return nil, fmt.Errorf("error parsing tenant key [%s]: %w", objects[i].Key, err)
+				}
+				if nextTenant != tenant {
+					break
+				}
 			}
 
 			if _, ok := tenants[tenant]; !ok {
-				tenants[tenant] = make([]string, 0, 100)
+				tenants[tenant] = nil // Initialize tenant with empty slice
```

Comment:
We do this since the function returns all tenants regardless of the filter. The filter filters out the files.
This way we can know which tenants are there for a given day, but filter the files whose retention are not expired yet.